src/map.rs

```diff
@@ -101,6 +101,64 @@ where
     out
 }
 
+/// Applies `f` to each subpixel of the input image in parallel.
+///
+/// The performance of this function depends on the dimensions of the image and the complexity of
+/// the mapping function `f`. For small images and fast mapping functions the non-parallel version
+/// is often faster whereas for large images with slow mapping functions the parallel version is
+/// often faster. It is a good idea to benchmark your individual usecases to make sure you are using the most
+/// appropriate version.
+///
+/// # Examples
+/// ```
+/// # extern crate image;
+/// # #[macro_use]
+/// # extern crate imageproc;
+/// # fn main() {
+/// use imageproc::map::map_subpixels_parallel;
+///
+/// let image = gray_image!(
+///     1, 2;
+///     3, 4);
+///
+/// let scaled = gray_image!(type: i16,
+///     -2, -4;
+///     -6, -8);
+///
+/// assert_pixels_eq!(
+///     map_subpixels_parallel(&image, |x| -2 * (x as i16)),
+///     scaled);
+/// # }
+/// ```
+#[cfg(feature = "rayon")]
+pub fn map_subpixels_parallel<I, P, F, S>(image: &I, f: F) -> Image<ChannelMap<P, S>>
+where
+    I: GenericImage<Pixel = P> + std::ops::Deref<Target = [P::Subpixel]>,
+    P: WithChannel<S>,
+    S: Primitive,
+    F: Fn(P::Subpixel) -> S + std::marker::Sync,
+    P::Subpixel: std::marker::Send,
+    P::Subpixel: std::marker::Sync,
+    P: std::marker::Send,
+    P: std::marker::Sync,
+    S: std::marker::Send,
+    S: std::marker::Sync,
+    <P as WithChannel<S>>::Pixel: std::marker::Send,
+    <P as WithChannel<S>>::Pixel: std::marker::Sync,
```

Comment:
This `where` clause is horrific! Is there any way of shortening it?

Reply:
we could add `use std::marker::{Send, Sync}` at the top of the file and then
```
    S: std::marker::Send,
    S: std::marker::Sync,
```
would become
```
    S: Send + Sync,
```

Reply:
the final result would be something like that : 
```
pub fn map_subpixels_parallel<I, P, F, S>(image: &I, f: F) -> Image<ChannelMap<P, S>>
where
    I: GenericImage<Pixel = P> + Deref<Target = [P::Subpixel]>,
    P: WithChannel<S> + Send + Sync,
    S: Primitive + Send + Sync,
    F: Fn(P::Subpixel) -> S + Sync,
    P::Subpixel: Send + Sync,
    <P as WithChannel<S>>::Pixel: Send + Sync,
```

Reply:
`Send` and `Sync` are already in rust prelude